tracing/src/main/java/com/palantir/tracing/CloseableTracer.java

```diff
@@ -25,10 +25,9 @@
  *
  * <p>Usage: try (CloseableTracer trace = CloseableTracer.start("traceName")) { [...] }
  */
-public class CloseableTracer implements AutoCloseable {
-    private static final CloseableTracer INSTANCE = new CloseableTracer();
+public abstract class CloseableTracer implements AutoCloseable {
```

Comment:
Would it be an ABI break to change this to an interface if there are no external implementations?

Reply:
The [JLS section on binary compatibility](https://docs.oracle.com/javase/specs/jls/se7/html/jls-13.html) doesn't seem to make any guarantees here, but https://stackoverflow.com/q/42294217 seems to suggest that this will work.

If we agree, I'd like to change this to an interface.

Reply:
I think [JLS section 13.4.16](https://docs.oracle.com/javase/specs/jls/se7/html/jls-13.html#jls-13.4.16) would be most applicable here.

The case I would be worried about are services using libraries compiled against an older `tracing-java` release where this is not an abstract class or interface, and where `tracing-java` version is newer and is abstract or interface leading to a runtime `LinkageError` (e.g. `AbstractMethodError`, `NoSuchMethodError`, or `IncompatibleClassChangeError`).

@carterkozak for thoughts on if this is safe.

Reply:
iirc making an abstract class concrete is usually fine, but swapping abstract class to/from an interface breaks ABI. I'm not certain off the top of my head about making a concrete method abstract (`close`). It would almost certainly be safe to provide an empty non-abstract implementation of the method, overridden in all implementations.

Reply:
Would you prefer if I did that instead of making the class abstract?